src/reflect/scala/reflect/internal/util/OwnerOnlyChmod.scala

```diff
@@ -0,0 +1,107 @@
+/* NSC -- new Scala compiler
+ * Copyright 2017 LAMP/EPFL
+ * @author  Martin Odersky
+ */
+package scala.reflect.internal.util
+
+import java.io.{File, FileOutputStream, IOException}
+
+
+trait OwnerOnlyChmod {
+  /** Remove group/other permissions for `file`, it if exists */
+  def chmod(file: java.io.File): Unit
+
+  /** Delete `file` if it exists, recreate it with no group/other permissions, and write `contents` */
+  final def chmodAndWrite(file: File, contents: Array[Byte]): Unit = {
+    file.delete()
+    val fos = new FileOutputStream(file)
+    fos.close()
+    chmod(file)
+    val fos2 = new FileOutputStream(file)
+    try {
+      fos2.write(contents)
+    } finally {
+      fos2.close()
+    }
+  }
```

Comment:
Note that creating a file _then_ changing the permissions on the file allows a race condition where an attacker on the system can open the file between the file create and the permissions change. It is far safer to use open(2) to set the permissions restrictive while creating the file.

Furthermore if this code is ever run on a directory with shared writing permissions, it's possible for an attacker to unlink(file), set a symlink in place, and have the process change modes of any target that the process has privileges to modify. fchmod(2) is the safe way to change permissions on a file that is already open, though that won't fix the above race condition. So consider this just an aside.

Thanks

Reply:
Thank you. Could you suggest how to do this on Java 6? I agree the race condition is a security hole, but couldn't figure out how to do this atomically with the API we have available. I think our NIO-based implementation on Java 8 is closer to what you're suggesting: https://github.com/scala/scala/pull/6120/files#diff-3578ac76088d22b5d9912e984dee3affR46.

Reply:
Thank you for the review.

We implemented things that way (or at least, tried to) in the 2.12 branch where we could link against the Java 8 NIO API. Scala 2.11 still supports Java 6.

https://github.com/scala/scala/blob/2.12.x/src/reflect/scala/reflect/internal/util/OwnerOnlyChmod.scala

Reply:
> adriaanm a minute ago
> retronym 29 seconds ago  

Speaking of race conditions :)

Reply:
The #diff-3578ac7.. link above looks like it may have the same race condition.

I don't know Java particularly well any longer, so it's hard to suggest a replacement. Two possibilities come to mind:

- set the umask to 0077 (octal), create the file, set the modes, and then reset umask to whatever the user wanted
- chmod 0700 (octal) the containing directory, create the file, set the file modes, then reset the directory to whatever permissions it should have.

I hope this helps. Thanks.

Reply:
How about: create the file with some temporary name/location, do what you need to do and then move it. file moves are atomic.